### PR TITLE
Multithreaded decoding

### LINKED ISSUE #21: Multithreaded decoding
It was once supported in the old versions, but it was removed since the implementation was too complicated to move forward.

We might need to support custom thread pool to increase interoperability.

### PR DESCRIPTION
**Pull request by tirr-c**
Closes #21 

This PR enables multithreaded decoding at various places in jxl-oxide. Specifically it adds:
- groupwise parallelism
- parallel Modular unsqueeze
- parallel edge-preserving filter
- frame parallelism (animation frames can be decoded in parallel)

Multithreading support is powered by Rayon and feature gated under `rayon` in `jxl-oxide` (`jxl-oxide-cli`), `jxl-render`, `jxl-modular` and a new crate, `jxl-threadpool`. `rayon` is enabled by default on `jxl-oxide`; you can disable default features to get a single threaded build.

### CODE REVIEW

File: crates/jxl-oxide/src/lib.rs [hunk 1]
```diff
@@ -106,9 +107,18 @@ pub use jxl_image::{ExtraChannelType, ImageHeader};
...
 
 pub type Result<T> = std::result::Result<T, Box<dyn std::error::Error + Send + Sync + 'static>>;
 
+fn default_pool() -> JxlThreadPool {
+    if cfg!(feature = "rayon") {
+        JxlThreadPool::default()
+    } else {
+        JxlThreadPool::none()
+    }
+}
```
Comment (EugeneVert) [inline crates/jxl-oxide/src/lib.rs:120]: `Default` for `JxlThreadPool` is implemented for for both `#[cfg(feature = "rayon")]` and `#[cfg(not(feature = "rayon"))]` cases

**Comment by EugeneVert** [comment]
Found a failing test case for cropping
```log
---- sunset_logo stdout ----
  Crop region: CropInfo { width: 409, height: 679, left: 429, top: 705 }
Testing frame #0
  Crop region: CropInfo { width: 264, height: 506, left: 491, top: 273 }
Testing frame #0
  Crop region: CropInfo { width: 179, height: 258, left: 527, top: 298 }
Testing frame #0
thread 'sunset_logo' panicked at 'assertion failed: (expected - actual).abs() <= 1e-6', crates/jxl-oxide/tests/crop.rs:51:25
```

File: crates/jxl-oxide/src/lib.rs (hunk 1 above)
Comment (tirr-c) [reply crates/jxl-oxide/src/lib.rs:120]: This is intended. I've wrote some explanation in the corresponding commit:

> Cargo features are additive, so thread pool support might be enabled
> even if it's disabled in jxl-oxide crate.

Well maybe I'm going to copy that commit message there...

**Comment by tirr-c** [comment]
Still cannot reproduce the crop test failure; I'll check that in another PR.

**Comment by tirr-c** [comment]
Cropped rendering is fixed, thread pool interface is changed to not providing `Default` implementation. Tests seems okay, merging!

**Pull request merged**